### ISSUE TITLE
Subcenter Arcface not dropping non-dominant class centers?

### ISSUE THREAD

**Issue by SNeugber** [KevinMusgrave/pytorch-metric-learning#740]
This is more of a question to check if I'm missing something, hope that's ok 🙏 

Shouldn't the method [get_outliers](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/src/pytorch_metric_learning/losses/subcenter_arcface_loss.py#L29) in the `SubCenterArcFaceLoss` class also be dropping the samples that don't belong to the dominiant class center? According to the end of section 3.2 from [the paper](https://arxiv.org/pdf/1801.07698):

> Based on the above observations, we propose a straightforward
approach to recapture intra-class compactness. We directly drop
non-dominant sub-centers after the network has enough discrimi-
native power. Meanwhile, we introduce a constant angle threshold
to drop high-confident noisy data. After that, we retrain the
ArcFace model from scratch on the automatically cleaned dataset.

Based on my understanding, Line 59 should therefore be:

```diff
- drop_dists = dominant_dist < cos_threshold
+ drop_dists = torch.logical_or(dominant_dist < cos_threshold, max_sub_center_idxs != dominant_idx)
```

On the other hand, the [official implementation](https://github.com/deepinsight/insightface/blob/master/recognition/subcenter_arcface/drop.py#L177) also only seems to keep samples based on the cosine similarity w.r.t. the dominant center.

Couldn't it be the case though that samples have a similarity to the dominant class center that is higher than the threshold, but still not as high as the similarity to the non-dominant class?

Thanks!

**Comment by SNeugber** [comment]
Nevermind, I think I got it...

I ended up plotting the distances between the samples and the dominant class center as well as the other class center (if the sample was closer to a non-dominant center) and saw that just rejecting based on cosine distance to the dominant center is fine 👍

The way I understand the paper now is that they drop the non-dominant classes _from the weights_, and only reject samples based on cosine distance w.r.t. the dominant class.

![Image](https://github.com/user-attachments/assets/58922ef0-c941-4221-b7a2-b4e67f61ec53)

**Comment by KevinMusgrave** [comment]
I appreciate the follow up!
